bundles/org.openhab.core.io.http.auth.basic/src/main/java/org/openhab/core/io/http/auth/basic/internal/BasicCredentialsExtractor.java

```diff
@@ -29,6 +30,7 @@
  */
 @Component(property = { "context=javax.servlet.http.HttpServletRequest" })
 public class BasicCredentialsExtractor implements CredentialsExtractor<HttpServletRequest> {
+    static HashMap<String, UsernamePasswordCredentials> authCache = new HashMap<String, UsernamePasswordCredentials>();
```

Comment:
It's very unsecure to cache plain text credentials. If you inspect a heap dump you can easily get hold of them. With a bit of creativity and reflection you might even be able to obtain them when you are able to run some malicious code/rules.

Reply:
It's base64 encoded so at least it's not directly readable. The OAuth2 tokens are cached, too. So whilst they don't leak the password it's possible to obtain them in the same way.

What would be another way to implement a cache?

Reply:
The best practices that I know of are to immediately hash it and compare it to the stored hash. After that you dispose the sensitive data ASAP to prevent it from leaking in any way. I guess the hashing is the part giving performance issues? The  tokens need to be refreshed every once in a while using the password. So if you have a token it is only valid until it expires which is more secure than having a password that can be used indefinitely. The data in users.json is also hashed so you cannot obtain the password from it.

Some users share heap dumps publicly to get some help without knowing that it contains this kind of sensitive data. You can get malicious code to run by sharing some .jar in the community forums or in the market place. If users use the same credentials with other services they are in even bigger troubles when someone gets hold of it.

Reply:
> I guess the hashing is the part giving performance issues?

Calculating hashes should really be a very quick and easy operation. I also hence wonder what exactly is causing the performance issues, then it can be better discussed what to do about it.

Reply:
> The tokens need to be refreshed every once in a while using the password or refresh token. So if you have a token it is only valid until it expires which is more secure than having a password that can be used indefinitely.

That is correct, but not for long living tokens because they never expire.

> If users use the same credentials with other services they are in even bigger troubles when someone gets hold of it.

Whilst that is correct that's imho a bit out of scope. Password reuse is very bad and everybody knows it.
I don't see the need to ensure this programatically.

Reply:
> I changed it so plaintext is no longer stored.

They are still stored unsecure because they are part of `UsernamePasswordCredentials`.

Reply:
> 
> They are still stored unsecure because they are part of `UsernamePasswordCredentials`.

Imho they contain the hashed password (with 65k iterations), not the plaintext password.